### PR TITLE
Completely eliminate casting navigation from Graql

### PR DESCRIPTION
**Pull request by aelred**

### CODE REVIEW

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java [hunk 1]
```diff
@@ -191,4 +167,33 @@ static void applyTypeLabelsToTraversal(
...
+        roleType.ifPresent(var -> {
+            // Access role-type ID from edge
+            Var roleTypeIdProperty = Var.anon();
+            Var edge = Var.anon();
+            traversal.as(edge.getValue()).values(ROLE_TYPE_ID.name()).as(roleTypeIdProperty.getValue());
+
+            // Look up direct role-type using ID
+            Var directRoleType = Var.anon();
+            GraphTraversal<Vertex, Vertex> vertexTraversal =
+                    traversal.V().as(directRoleType.getValue())
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java]: I have no doubt that this works but does it perform? Have you confirmed the index is being used?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/InIsaFragment.java [hunk 2]
```diff
@@ -30,48 +30,24 @@
...
     }
 
     @Override
     public void applyTraversal(GraphTraversal<Vertex, Vertex> traversal, GraknGraph graph) {
         Fragments.inSubs(traversal).in(SHARD.getLabel()).in(ISA.getLabel());
-        if (!allowCastings) {
-            // Make sure we never get any castings
-            traversal.not(__.hasLabel(Schema.BaseType.CASTING.name()));
-        }
+        // Make sure we never get any castings
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/InIsaFragment.java]: Almost there on not having to do this :)

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/OutShortcutFragment.java [hunk 3]
```diff
@@ -46,36 +47,45 @@
...
+    private final Optional<Set<TypeLabel>> roleTypeLabels;
+    private final Optional<Set<TypeLabel>> relationTypeLabels;
 
     OutShortcutFragment(
-            Var relation, Var edge, Var rolePlayer, Optional<Set<TypeLabel>> roleTypes,
-            Optional<Set<TypeLabel>> relationTypes) {
-            super(relation, rolePlayer, edge);
+            Var relation, Var edge, Var rolePlayer, Optional<Var> roleType, Optional<Set<TypeLabel>> roleTypeLabels,
+            Optional<Set<TypeLabel>> relationTypeLabels) {
+            super(relation, rolePlayer, roleType.map(rt -> new Var[] {edge, rt}).orElse(new Var[] {edge}));
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/OutShortcutFragment.java]: I would have gone with creating a new constructor. This is horribly difficult to read. Just a suggestion though.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java (hunk 1 above)
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java]: I don't know. I'll investigate - don't think I'll be able to write a test for it though!

**Comment by aelred** [comment]
@fppt humble re-review request

File: grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanInternalFactory.java [hunk 4]
```diff
@@ -88,6 +94,14 @@ private synchronized TitanGraph newTitanGraph(String name, String address, Prope
         TitanGraph titanGraph = configureGraph(name, address, properties, batchLoading);
         buildTitanIndexes(titanGraph);
         titanGraph.tx().onClose(Transaction.CLOSE_BEHAVIOR.ROLLBACK);
+
+        if (!strategiesApplied.getAndSet(true)) {
```
Comment (fppt) [inline grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanInternalFactory.java]: Nice compromise

File: grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStep.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    TitanPreviousPropertyStep(Traversal.Admin traversal, String propertyKey, String stepLabel) {
+        super(traversal);
+        this.propertyKey = Objects.requireNonNull(propertyKey);
+        this.stepLabel = Objects.requireNonNull(stepLabel);
+    }
+
+    @Override
+    protected Iterator<TitanVertex> flatMap(Traverser.Admin<S> traverser) {
+        TitanTransaction tx = TitanTraversalUtil.getTx(this.traversal);
+        Object value = getNullableScopeValue(Pop.first, stepLabel, traverser);
```
Comment (fppt) [inline grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStep.java]: I think this still needs some clarrifying. What value is being extracted here?

File: grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStep.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @Override
+    public Set<String> getScopeKeys() {
+        return Collections.singleton(stepLabel);
+    }
+
+    @Override
+    public Set<TraverserRequirement> getRequirements() {
+        return TYPICAL_GLOBAL_REQUIREMENTS;
```
Comment (fppt) [inline grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStep.java]: Magic

File: grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStepStrategy.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    private static final long serialVersionUID = 6888929702831948298L;
+
+    @Override
+    public void apply(Traversal.Admin<?, ?> traversal) {
+
+        // Retrieve all graph (`V()`) steps - this is the step the strategy should replace
+        List<TitanGraphStep> graphSteps = TraversalHelper.getStepsOfClass(TitanGraphStep.class, traversal);
+
+        for (TitanGraphStep graphStep : graphSteps) {
+            // For each graph step, confirm it follows this pattern:
```
Comment (fppt) [inline grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStepStrategy.java]: YESSS much clearer

File: grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStepStrategy.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        return filterStep.getLocalChildren().get(0).getSteps();
+    }
+
+    private Optional<String> propertyFromPropertiesStep(PropertiesStep<Vertex> propertiesStep) {
+        String[] propertyKeys = propertiesStep.getPropertyKeys();
+        if (propertyKeys.length != 1) return Optional.empty();
+        return Optional.of(propertyKeys[0]);
+    }
+
+    private Optional<String> eqPredicateFromWherePredicate(WherePredicateStep<Vertex> whereStep) {
```
Comment (fppt) [inline grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStepStrategy.java]: maybe this method can be called `getLabelFromEqualsPredicate` I think that makes more sense.

**Comment by aelred** [comment]
@fppt re-re-review 👍

**Comment by aelred** [comment]
!rtg - ai.grakn.test.engine.controller.TasksControllerTest.afterSendingTaskWithRunAt_ItIsDelayedInStorage

**Pull request merged**